File: ts/src/signing-keys.ts
/**
 * Owner-facing signing-key surface (roles/signing-keys overhaul).
 *
 * A platform owner reads their realm's keyring and self-serve rotates the
 * active signing key:
 *   - GET  /platforms/{id}/signing-keys        — keyring + rotation policy
 *   - POST /platforms/{id}/signing-keys/rotate — mint a new active key
 *
 * Distinct from the base-staff ops rotate at `/admin/platforms/{id}/…`
 * (that surface lives only in `@realm-id/web-admin`). Both are realm-admin
 * gated server-side; this client targets the owner's own realm.
 */

import type { HttpClient } from "./http.js";

/** One key in the realm's signing keyring. */
export interface SigningKey {
  kid: string;
  /** Unix seconds the key was created. */
  created_at: number;
  /** Unix seconds the key stops being used to sign new tokens. */
  active_until: number;
  /** Unix seconds the key is dropped from the JWKS entirely. */
  retire_at: number;
  /** True for the key currently minting tokens. */
  is_current: boolean;
}

/** The realm's rotation policy as reported by the keyring read. */
export interface SigningKeyRotation {
  mode: "auto" | "manual";
  /** Cadence when `mode === "auto"`; omitted for manual or unset. */
  interval?: "1w" | "1mo" | "1y";
  /** Unix seconds the worker next mints a replacement; omitted in manual mode. */
  next_rotation_at?: number;
}

export interface SigningKeysResponse {
  keys: SigningKey[];
  rotation: SigningKeyRotation;
}

/** Result of a rotate: the new current `kid` plus any retired kids. */
export interface RotateSigningKeyResult {
  kid: string;
  retired_kids: string[];
}

export class SigningKeysClient {
  constructor(
    private readonly http: HttpClient,
    private readonly realmId: string,
  ) {}

  /** GET /platforms/{id}/signing-keys — keyring newest-first + rotation policy. */
  async list(): Promise<SigningKeysResponse> {
    const raw = await this.http.request<SigningKeysResponse>({
      method: "GET",
      path: `/platforms/${encodeURIComponent(this.realmId)}/signing-keys`,
    });
    return {
      keys: Array.isArray(raw?.keys) ? raw.keys : [],
      rotation: raw?.rotation ?? { mode: "auto" },
    };
  }

  /**
   * POST /platforms/{id}/signing-keys/rotate — self-serve rotate. Shares the
   * server-side rotator + rate limiter with the ops route (429 `rate_limited`
   * when called too frequently).
   */
  async rotate(): Promise<RotateSigningKeyResult> {
    const raw = await this.http.request<RotateSigningKeyResult>({
      method: "POST",
      path: `/platforms/${encodeURIComponent(this.realmId)}/signing-keys/rotate`,
    });
    return { kid: raw?.kid ?? "", retired_kids: raw?.retired_kids ?? [] };
  }
}

File: go/signing_keys.go
package realmid

import (
	ctxpkg "context"
	"net/url"
)

// Owner-facing signing-key surface (roles/signing-keys overhaul).
//
// A platform owner reads their realm's keyring and self-serve rotates the
// active signing key:
//   - GET  /platforms/{id}/signing-keys        — keyring + rotation policy
//   - POST /platforms/{id}/signing-keys/rotate — mint a new active key
//
// Distinct from the base-staff ops rotate at `/admin/platforms/{id}/…`
// (not part of this partner SDK). Both are realm-admin gated server-side;
// this client targets the caller's own realm.

// SigningKey is one key in the realm's signing keyring.
type SigningKey struct {
	KID string `json:"kid"`
	// CreatedAt is the unix-seconds timestamp the key was created.
	CreatedAt int64 `json:"created_at"`
	// ActiveUntil is when the key stops signing new tokens (unix seconds).
	ActiveUntil int64 `json:"active_until"`
	// RetireAt is when the key drops out of the JWKS entirely (unix seconds).
	RetireAt int64 `json:"retire_at"`
	// IsCurrent is true for the key currently minting tokens.
	IsCurrent bool `json:"is_current"`
}

// SigningKeyRotation is the realm's rotation policy.
type SigningKeyRotation struct {
	// Mode is "auto" or "manual".
	Mode string `json:"mode"`
	// Interval is the cadence ("1w"/"1mo"/"1y") when Mode=="auto"; empty
	// for manual or unset.
	Interval string `json:"interval,omitempty"`
	// NextRotationAt is when the worker next mints a replacement (unix
	// seconds); zero in manual mode.
	NextRotationAt int64 `json:"next_rotation_at,omitempty"`
}

// SigningKeysResponse is the GET /platforms/{id}/signing-keys body.
type SigningKeysResponse struct {
	Keys     []SigningKey       `json:"keys"`
	Rotation SigningKeyRotation `json:"rotation"`
}

// RotateSigningKeyResult is the rotate acknowledgment: the new current
// KID plus any KIDs retired by the rotation.
type RotateSigningKeyResult struct {
	KID         string   `json:"kid"`
	RetiredKIDs []string `json:"retired_kids"`
}

// SigningKeysClient is realm.SigningKeys.
type SigningKeysClient struct {
	realm *Realm
}

// List reads the realm's keyring (newest-first) and rotation policy.
func (c *SigningKeysClient) List(ctx ctxpkg.Context) (*SigningKeysResponse, error) {
	tok, err := c.realm.platformToken.get(ctx)
	if err != nil {
		return nil, err
	}
	var out SigningKeysResponse
	if err := c.realm.http.do(ctx, requestOptions{
		Method: "GET",
		Path:   "/platforms/" + url.PathEscape(c.realm.realmID) + "/signing-keys",
		Bearer: tok,
	}, &out); err != nil {
		return nil, err
	}
	if out.Keys == nil {
		out.Keys = []SigningKey{}
	}
	return &out, nil
}

// Rotate self-serve rotates the realm's active signing key. Shares the
// server-side rotator + rate limiter with the ops route (a 429
// `rate_limited` RealmError is returned when called too frequently).
func (c *SigningKeysClient) Rotate(ctx ctxpkg.Context) (*RotateSigningKeyResult, error) {
	tok, err := c.realm.platformToken.get(ctx)
	if err != nil {
		return nil, err
	}
	var out RotateSigningKeyResult
	if err := c.realm.http.do(ctx, requestOptions{
		Method: "POST",
		Path:   "/platforms/" + url.PathEscape(c.realm.realmID) + "/signing-keys/rotate",
		Bearer: tok,
	}, &out); err != nil {
		return nil, err
	}
	if out.RetiredKIDs == nil {
		out.RetiredKIDs = []string{}
	}
	return &out, nil
}
